src/python/pants/util/objects.py

```diff
@@ -281,6 +284,20 @@ def __new__(cls, value):
       """
       return cls.create(value)
 
+    # TODO: figure out if this will always trigger on primitives like strings, and what situations
+    # won't call this __eq__ (and therefore won't raise like we want).
+    def __eq__(self, other):
+      """Redefine equality to raise to nudge people to use static pattern matching."""
+      raise self.make_type_error(
```

Comment:
I would rather only raise if the other isn't of the same type. I think it's totally valid to write `toolchain_variant != ToolchainVariant.gnu`, and that that is cleaner than the nested match you have in the ctypes test.